java/query/query-lambda/src/main/java/sleeper/query/lambda/WarmQueryExecutorLambda.java

```diff
@@ -0,0 +1,149 @@
+ * The results from the queries are discarded as they are not required since this Lambda ensures that the query Lambdas
+ * remain warm for genuine queries.
+ */
+@SuppressWarnings("unused")
+public class WarmQueryExecutorLambda implements RequestHandler<ScheduledEvent, Void> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(WarmQueryExecutorLambda.class);
+
+    private Instant lastUpdateTime;
+    private InstanceProperties instanceProperties;
+    private TablePropertiesProvider tablePropertiesProvider;
+    private final AmazonSQS sqsClient;
+    private final AmazonS3 s3Client;
+    private final AmazonDynamoDB dynamoClient;
+    private QueryMessageHandler messageHandler;
+    private SqsQueryProcessor processor;
+    private final DynamoDBQueryTracker queryTracker;
+
+    public WarmQueryExecutorLambda() throws ObjectFactoryException {
+        this(AmazonS3ClientBuilder.defaultClient(), AmazonSQSClientBuilder.defaultClient(),
+                AmazonDynamoDBClientBuilder.defaultClient(), System.getenv(CONFIG_BUCKET.toEnvironmentVariable()));
+    }
+
+    public WarmQueryExecutorLambda(AmazonS3 s3Client, AmazonSQS sqsClient, AmazonDynamoDB dynamoClient, String configBucket) throws ObjectFactoryException {
+        this.s3Client = s3Client;
+        this.sqsClient = sqsClient;
+        this.dynamoClient = dynamoClient;
+        instanceProperties = loadInstanceProperties(s3Client, configBucket);
+        tablePropertiesProvider = new TablePropertiesProvider(instanceProperties, s3Client, dynamoClient);
+        queryTracker = new DynamoDBQueryTracker(instanceProperties, dynamoClient);
+    }
+
+    @Override
+    public Void handleRequest(ScheduledEvent event, Context context) {
+        LOGGER.info("Starting to build queries for the tables");
+        S3TableProperties.getStore(instanceProperties, s3Client, dynamoClient)
+                .streamAllTables()
+                .forEach(tableProperty -> {
+                    Schema schema = tableProperty.getSchema();
+                    Field field = schema.getRowKeyFields().get(0);
+
+                    // Create the max and min values for the query. We don't care what they are as long as the query runs
+                    Type type = field.getType();
+                    Object min, max;
+                    if (type instanceof IntType) {
+                        min = 0;
+                        max = 1;
+                    } else if (type instanceof LongType) {
+                        min = 0L;
+                        max = 1L;
+                    } else if (type instanceof StringType) {
+                        min = "a";
+                        max = min;
+                    } else if (type instanceof ByteArrayType) {
+                        min = new byte[]{'a'};
+                        max = min;
+                    } else {
+                        throw new IllegalArgumentException("Unknown type in the schema: " + type);
+                    }
```

Comment:
This could still cover a lot of data. If there's more than one row key, there could be a very large number of records all with the exact same value for the first row key. If sort keys are used, there could be a large number of records all with the same value for all of the row keys.

We could create a region with an exact value for each row key and sort key.

You could extract the code for generating a query into a separate method that takes the TableProperties and a query ID, and returns a Query. If you write unit tests against that method, rather than integration tests, the tests would run much faster and it should make it easier for you to test it.

It's still nice to have an integration test like you've got as well, but if you write unit tests you could make it just one integration test.

Reply:
Updated a new tests added